# Анализ данных о компьютерных играх

## Описание исследования

В распоряжении имеются данные о компьютерных играх: название игры, платформа, год выпуска, жанр, продажи в разных регионах, оценки критиков и пользователей, а также возрастной рейтинг.

Эти данные позволяют изучить структуру игрового рынка, выявить особенности игр, выпущенных в разные годы и на разных платформах, а также подготовить данные для дальнейшего анализа.

## Цель исследования

Цель исследования — подготовить данные о компьютерных играх для дальнейшего анализа и изучить основные характеристики игр за актуальный период.

В рамках работы необходимо обработать пропуски, исправить ошибки в данных, удалить дубликаты, выделить актуальный период и определить наиболее представленные платформы по количеству выпущенных игр.

## Задачи исследования

Для достижения цели необходимо:

1. Изучить общую информацию о датасете.
2. Привести названия столбцов к единому стилю.
3. Проверить типы данных и при необходимости преобразовать их.
4. Изучить пропущенные значения и принять решение об их обработке.
5. Проверить данные на наличие явных и неявных дубликатов.
6. Исправить обнаруженные ошибки и аномалии.
7. Выделить актуальный период для дальнейшего анализа.
8. Провести категоризацию оценок пользователей и критиков.
9. Определить топ-7 платформ по количеству выпущенных игр за актуальный период.
10. Сформулировать общий вывод по результатам подготовки и первичного анализа данных.

## Исходные данные

В исследовании используется таблица с данными о компьютерных играх.

Описание столбцов:

* `name` — название игры;
* `platform` — игровая платформа;
* `year_of_release` — год выпуска игры;
* `genre` — жанр игры;
* `na_sales` — продажи в Северной Америке;
* `eu_sales` — продажи в Европе;
* `jp_sales` — продажи в Японии;
* `other_sales` — продажи в других регионах;
* `critic_score` — оценка критиков;
* `user_score` — оценка пользователей;
* `rating` — возрастной рейтинг игры.

## План исследования

Данное исследование разделим на несколько частей.

### Часть 1. Изучение общей информации

1. Загрузка библиотек.
2. Загрузка файла с данными.
3. Изучение структуры таблицы.
4. Просмотр общей информации о данных.

### Часть 2. Подготовка данных

1. Приведение названий столбцов к нижнему регистру и стилю `snake_case`.
2. Преобразование данных к нужным типам.
3. Изучение и обработка пропусков.
4. Проверка уникальных значений в категориальных столбцах.
5. Исправление неявных дубликатов.
6. Проверка и удаление явных дубликатов.

### Часть 3. Анализ актуального периода

1. Выделение актуального периода с 2000 по 2013 год включительно.
2. Создание отдельного датафрейма `df_actual`.
3. Категоризация оценок пользователей и критиков с помощью пользовательских функций.
4. Выделение топ-7 платформ по количеству выпущенных игр за актуальный период.

### Часть 4. Общий вывод

1. Обобщение результатов предобработки данных.
2. Описание принятых решений по пропускам, типам данных и дубликатам.
3. Формулировка выводов по актуальному периоду и наиболее представленным платформам.


In [1]:
import pandas as pd


In [2]:
data = pd.read_csv('new_games.csv')

In [3]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
data.columns = data.columns.str.lower().str.replace(' ' , '_').str.replace('-', '_')

In [5]:
print(data.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


In [6]:
data['eu_sales'] = pd.to_numeric(data['eu_sales'] , errors='coerce')

In [7]:
data['na_sales'] = pd.to_numeric(data['na_sales'] , errors='coerce')

In [8]:
data['jp_sales'] = pd.to_numeric(data['jp_sales'] , errors='coerce')

In [9]:
data['year_of_release'] = pd.to_numeric(data['year_of_release'] , errors='coerce')

In [10]:
data['year_of_release'] = data['year_of_release'].astype('Int64')

In [11]:
data['rating'] = data['rating'].fillna('unknown')

In [12]:
data['critic_score'] = pd.to_numeric(data['critic_score'] , errors='coerce')

In [13]:
data['other_sales'] = pd.to_numeric(data['other_sales'] , errors='coerce')

In [14]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  Int64  
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           16956 non-null  object 
dtypes: Int64(1), float64(5), object(5)
memory usage: 1.4+ MB
None


Названия столбцов были приведены к нижнему регистру и стилю snake_case. 
Столбцы с продажами, оценками пользователей и критиков были приведены к числовому типу. 
Некорректные строковые значения при преобразовании были заменены на пропуски. 
Пропуски в critic_score и user_score не заполнялись, так как оценки невозможно корректно восстановить по имеющимся данным. 
Пропуски в rating были заменены на unknown, поскольку рейтинг является категориальным признаком, а неизвестные значения можно выделить в отдельную категорию.

In [15]:
data.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         6804
rating                0
dtype: int64

In [16]:
data.isna().mean() * 100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         40.127389
rating              0.000000
dtype: float64

После обработки данных пропуски остались в основном в столбцах critic_score и user_score: около 51.39% и 40.13% соответственно. Эти значения не заполнялись, так как оценки критиков и пользователей невозможно корректно восстановить по имеющимся данным без искажения анализа.
В year_of_release осталось около 1.62% пропусков. В столбцах name, genre, eu_sales, jp_sales доля пропусков незначительная.
Пропуски в rating были заменены на значение unknown, поэтому после обработки в этом столбце отсутствуют значения NaN.

Наибольшая доля пропусков наблюдается в столбцах critic_score и user_score. 
Пропуски в user_score могут быть связаны с тем, что не все пользователи оставляют оценки играм, особенно если игра была малопопулярной, старой или выходила на менее массовой платформе.

Пропуски в critic_score, вероятно, объясняются тем, что профессиональные критики оценивают не все игры. Больше внимания обычно получают крупные и заметные релизы, тогда как нишевые, старые или региональные проекты могли не иметь достаточного количества рецензий. Поэтому отсутствие оценки критиков не стоит заменять средним или медианным значением, так как это исказило бы реальные данные.

В столбцах с региональными продажами были обнаружены единичные пропуски. Их доля крайне мала относительно общего объёма датасета, поэтому было принято решение не удалять эти строки. Заполнять такие значения средним или медианным значением некорректно, так как продажи игр распределены неравномерно: небольшое количество популярных игр может иметь очень высокие продажи и заметно искажать средние показатели. Поэтому пропуски в столбцах с продажами были оставлены без изменения, чтобы не вносить искусственные значения в данные.


In [17]:
data[['na_sales', 'eu_sales', 'jp_sales', 'other_sales']].isna().sum()

na_sales       0
eu_sales       6
jp_sales       4
other_sales    0
dtype: int64

In [18]:
columns = ['genre', 'platform', 'rating', 'year_of_release']

for column in columns:
    print(column)
    print(data[column].unique())
    print()

genre
['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' nan 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']

platform
['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX']

rating
['E' 'unknown' 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']

year_of_release
<IntegerArray>
[2006, 1985, 2008, 2009, 1996, 1989, 1984, 2005, 1999, 2007, 2010, 2013, 2004,
 1990, 1988, 2002, 2001, 2011, 1998, 2015, 2012, 2014, 1992, 1997, 1993, 1994,
 1982, 2016, 2003, 1986, 2000, <NA>, 1995, 1991, 1981, 1987, 1980, 1983]
Length: 38, dtype: Int64



In [19]:
data['genre'] = data['genre'].str.lower().str.strip()

In [20]:
print(data['genre'].unique)

<bound method Series.unique of 0              sports
1            platform
2              racing
3              sports
4        role-playing
             ...     
16951          action
16952          sports
16953       adventure
16954        platform
16955      simulation
Name: genre, Length: 16956, dtype: object>


В столбце genre были обнаружены неявные дубликаты, связанные с разным регистром написания одних и тех же жанров: например, Sports и SPORTS, Platform и PLATFORM. Для устранения этой проблемы значения в столбце genre были приведены к нижнему регистру. После повторной проверки уникальных значений неявные дубликаты в genre исчезли.

В столбцах platform, rating и year_of_release неявных дубликатов не обнаружено. В rating значение unknown отражает игры с неизвестным возрастным рейтингом, а в year_of_release сохраняются отдельные пропущенные значения.

In [21]:
data.duplicated


<bound method DataFrame.duplicated of                                 name platform  year_of_release         genre  \
0                         Wii Sports      Wii             2006        sports   
1                  Super Mario Bros.      NES             1985      platform   
2                     Mario Kart Wii      Wii             2008        racing   
3                  Wii Sports Resort      Wii             2009        sports   
4           Pokemon Red/Pokemon Blue       GB             1996  role-playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3             2016        action   
16952               LMA Manager 2007     X360             2006        sports   
16953        Haitaka no Psychedelica      PSV             2016     adventure   
16954               Spirits & Spells      GBA             2003      platform   
16955            Winning Post 8 2016      PSV             2016    simulation   

 

In [22]:
data.loc[data['rating'] != 'unknown'].duplicated().sum()

np.int64(136)

In [23]:
data.duplicated().sum()

np.int64(241)

In [24]:
data.loc[data.duplicated(keep=False)]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
267,Batman: Arkham Asylum,PS3,2009,action,2.24,1.31,0.07,0.61,91.0,8.9,T
268,Batman: Arkham Asylum,PS3,2009,action,2.24,1.31,0.07,0.61,91.0,8.9,T
367,James Bond 007: Agent Under Fire,PS2,2001,shooter,1.90,1.13,0.10,0.41,72.0,7.9,T
368,James Bond 007: Agent Under Fire,PS2,2001,shooter,1.90,1.13,0.10,0.41,72.0,7.9,T
716,God of War: Ascension,PS3,2013,action,1.23,0.63,0.04,0.35,80.0,7.5,M
...,...,...,...,...,...,...,...,...,...,...,...
16799,Transformers: Prime,Wii,2012,action,0.00,0.01,0.00,0.00,NaN,NaN,unknown
16911,Metal Gear Solid V: The Definitive Experience,XOne,2016,action,0.01,0.00,0.00,0.00,NaN,tbd,M
16912,Metal Gear Solid V: The Definitive Experience,XOne,2016,action,0.01,0.00,0.00,0.00,NaN,tbd,M
16939,The Longest 5 Minutes,PSV,2016,action,0.00,0.00,0.01,0.00,NaN,NaN,unknown


In [25]:
data.loc[data['rating'] == 'unknown']

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
1,Super Mario Bros.,NES,1985,platform,29.08,3.58,6.81,0.77,NaN,NaN,unknown
4,Pokemon Red/Pokemon Blue,GB,1996,role-playing,11.27,8.89,10.22,1.00,NaN,NaN,unknown
5,Tetris,GB,1989,puzzle,23.20,2.26,4.22,0.58,NaN,NaN,unknown
9,Duck Hunt,NES,1984,shooter,26.93,0.63,0.28,0.47,NaN,NaN,unknown
10,Nintendogs,DS,2005,simulation,9.05,10.95,1.93,2.74,NaN,NaN,unknown
...,...,...,...,...,...,...,...,...,...,...,...
16951,Samurai Warriors: Sanada Maru,PS3,2016,action,0.00,0.00,0.01,0.00,NaN,NaN,unknown
16952,LMA Manager 2007,X360,2006,sports,0.00,0.01,0.00,0.00,NaN,NaN,unknown
16953,Haitaka no Psychedelica,PSV,2016,adventure,0.00,0.00,0.01,0.00,NaN,NaN,unknown
16954,Spirits & Spells,GBA,2003,platform,0.01,0.00,0.00,0.00,NaN,NaN,unknown


In [26]:
data.loc[data['rating'] == 'unknown'].duplicated().sum()

np.int64(105)

In [27]:
unknown_rating = data.loc[data['rating'] == 'unknown']

unknown_rating.loc[unknown_rating.duplicated(keep=False)]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
822,Wipeout: The Game,Wii,2009,misc,1.94,0.00,0.00,0.12,NaN,NaN,unknown
823,Wipeout: The Game,Wii,2009,misc,1.94,0.00,0.00,0.12,NaN,NaN,unknown
1242,NASCAR 99,PS,1998,racing,1.45,0.04,0.00,0.04,NaN,NaN,unknown
1243,NASCAR 99,PS,1998,racing,1.45,0.04,0.00,0.04,NaN,NaN,unknown
1298,LEGO City Undercover,3DS,2013,platform,0.52,0.77,0.07,0.11,NaN,NaN,unknown
...,...,...,...,...,...,...,...,...,...,...,...
16753,Routes PE,PS2,2007,adventure,0.00,0.00,0.01,0.00,NaN,NaN,unknown
16798,Transformers: Prime,Wii,2012,action,0.00,0.01,0.00,0.00,NaN,NaN,unknown
16799,Transformers: Prime,Wii,2012,action,0.00,0.01,0.00,0.00,NaN,NaN,unknown
16939,The Longest 5 Minutes,PSV,2016,action,0.00,0.00,0.01,0.00,NaN,NaN,unknown


In [28]:
data = data.drop_duplicates()

In [29]:
data.duplicated().sum()

np.int64(0)

Была выполнена проверка на наличие явных дубликатов. Всего в датасете было обнаружено 241 полностью повторяющаяся строка. Дополнительно были отдельно проверены строки со значением unknown в столбце rating, так как это значение было добавлено на этапе обработки пропусков. Среди строк с известным рейтингом было найдено 136 дубликатов, следовательно, часть повторов приходилась и на строки с unknown.

Поскольку обнаруженные дубликаты полностью повторяли уже имеющиеся записи, они могли искажать результаты анализа, например завышать количество игр и суммарные продажи. Поэтому 241 явная дублирующаяся строка была удалена. После удаления повторная проверка показала отсутствие явных дубликатов.

In [30]:
print(data['year_of_release']) 

0        2006
1        1985
2        2008
3        2009
4        1996
         ... 
16951    2016
16952    2006
16953    2016
16954    2003
16955    2016
Name: year_of_release, Length: 16715, dtype: Int64


In [31]:
df_actual = data.loc[(data['year_of_release'] >= 2000) & (data['year_of_release'] <= 2013)].copy()

In [32]:
df_actual['year_of_release'].min(), df_actual['year_of_release'].max()

(np.int64(2000), np.int64(2013))

In [33]:
print(data['user_score'])

0          8
1        NaN
2        8.3
3          8
4        NaN
        ... 
16951    NaN
16952    NaN
16953    NaN
16954    NaN
16955    NaN
Name: user_score, Length: 16715, dtype: object


In [34]:
def categorize_user_score(score):
    if pd.isna(score):
        return 'без рейтинга'
    elif 8 <= score <= 10:
        return 'высокая оценка'
    elif 3 <= score < 8:
        return 'средняя оценка'
    elif 0 <= score < 3:
        return 'низкая оценка'
    else:
        return 'некорректное значение'

In [35]:
def categorize_critic_score(score):
    if pd.isna(score):
        return 'без рейтинга'
    elif 80 <= score <= 100:
        return 'высокая оценка'
    elif 30 <= score < 80:
        return 'средняя оценка'
    elif 0 <= score < 30:
        return 'низкая оценка'
    else:
        return 'некорректное значение'

In [36]:
df_actual['user_score'] = pd.to_numeric(df_actual['user_score'], errors='coerce')

In [37]:
df_actual['categorize_user_score'] = df_actual['user_score'].apply(categorize_user_score)

In [38]:
df_actual['critic_score'] = pd.to_numeric(df_actual['critic_score'], errors='coerce')

In [39]:
df_actual['critic_score_category'] = df_actual['critic_score'].apply(categorize_critic_score)

In [40]:
top_platforms = df_actual['platform'].value_counts().head(7)

In [41]:
print(top_platforms)

platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: count, dtype: int64


Для анализа был выделен актуальный период с 2000 по 2013 год включительно. На его основе был создан отдельный датафрейм `df_actual`, чтобы дальнейшие расчёты проводились только по играм, выпущенным в выбранный период.

Также были созданы пользовательские функции для категоризации оценок пользователей и критиков. Для `user_score` использовалась шкала от 0 до 10, а для `critic_score` — шкала от 0 до 100. Оценки были разделены на категории: низкая, средняя, высокая, а пропущенные значения были вынесены в отдельную категорию «без рейтинга». Это позволяет анализировать оценки не только как числовые значения, но и как более удобные смысловые группы.

После этого были выделены топ-7 платформ по количеству выпущенных игр за актуальный период. Самыми представленными платформами оказались PS2, DS, Wii, PSP, X360, PS3 и GBA. Больше всего игр было выпущено на PS2 — 2154 игры, почти столько же на DS — 2146 игр. Далее идут Wii — 1294 игры, PSP — 1199 игр, X360 — 1138 игр, PS3 — 1107 игр и GBA — 826 игр.

Таким образом, в актуальном периоде наиболее популярными по количеству выпущенных игр стали платформы PS2 и DS. Это говорит о том, что именно они занимали наиболее заметное место в датасете по числу релизов за выбранные годы.
